File: files/300-399/363/CH-363.py
aaimport pandas as pd
import numpy as np

path = "300-399/363/CH-363 Matrix Calculation.xlsx"

input_data = pd.read_excel(path,header=None)
input_matrix = input_data.iloc[3:7, 2:7].to_numpy()
test1 = input_data.iloc[3, 9:12].to_numpy()
test2 = input_data.iloc[3, 14:19].to_numpy()
test3 = input_data.iloc[3:7, 22].to_numpy()

par = 16

def find_subvectors_sum_to_par(mat, target_sum):
    results = []
    for i in range(mat.shape[0]):
        for j in range(mat.shape[1]):
            for k in range(j, mat.shape[1]):
                subvector = mat[i, j:k+1]
                if np.sum(subvector) == target_sum:
                    results.append(subvector.flatten().tolist())
    for j in range(mat.shape[1]):
        for i in range(mat.shape[0]):
            for k in range(i, mat.shape[0]):
                subvector = mat[i:k+1, j]
                if np.sum(subvector) == target_sum:
                    results.append(subvector.flatten().tolist())

    return results

result = find_subvectors_sum_to_par(input_matrix, par)

print(np.all(result[0] == test1)) # True
print(np.all(result[1] == test2)) # True
print(np.all(result[2] == test3)) # True

File: files/300-399/363/CH-363.R
library(tidyverse)
library(readxl)

path <- "300-399/363/CH-363 Matrix Calculation.xlsx"
input <- read_excel(path, range = "C4:G7", col_names = FALSE) %>% as.matrix()
test1 <- read_excel(path, range = "J4:L4", col_names = FALSE) %>% as.matrix()
test2 <- read_excel(path, range = "O4:S4", col_names = FALSE) %>% as.matrix()
test3 <- read_excel(path, range = "W4:W7", col_names = FALSE) %>% as.matrix()

par = 16

find_subvectors_sum_to_par <- function(mat, target_sum) {
  results <- list()

  for (i in 1:nrow(mat)) {
    for (j in 1:ncol(mat)) {
      for (k in j:ncol(mat)) {
        subvector <- mat[i, j:k]
        if (sum(subvector) == target_sum) {
          results <- append(results, list(matrix(subvector, nrow = 1)))
        }
      }
    }
  }

  for (j in 1:ncol(mat)) {
    for (i in 1:nrow(mat)) {
      for (k in i:nrow(mat)) {
        subvector <- mat[i:k, j]
        if (sum(subvector) == target_sum) {
          results <- append(results, list(matrix(subvector, ncol = 1)))
        }
      }
    }
  }

  return(results)
}

result <- find_subvectors_sum_to_par(input, par)

all(result[[1]] == test1)
all(result[[2]] == test2)
all(result[[3]] == test3)
